jclouds-plugin/src/main/java/jenkins/plugins/jclouds/compute/JCloudsSlaveTemplate.java

```diff
@@ -312,41 +389,27 @@ public NodeMetadata get() {
         }
 
         Statement initStatement = null;
-        Statement bootstrap = null;
 
         if (this.preExistingJenkinsUser) {
             if (this.initScript.length() > 0) {
                 initStatement = Statements.exec(this.initScript);
             }
         } else {
-            // setup the jcloudTemplate to customize the nodeMetadata with jdk, etc. also opening ports
+            // provision jenkins user
             AdminAccess adminAccess = AdminAccess.builder().adminUsername(getJenkinsUser())
-                    .installAdminPrivateKey(installPrivateKey) // some VCS such as Git use SSH authentication
-                    .grantSudoToAdminUser(allowSudo) // no need
-                    .adminPrivateKey(getCloud().privateKey) // temporary due to jclouds bug
-                    .authorizeAdminPublicKey(true).adminPublicKey(getCloud().publicKey).adminHome(getFsRoot()).build();
-
+                .installAdminPrivateKey(installPrivateKey) // some VCS such as Git use SSH authentication
+                .grantSudoToAdminUser(allowSudo) // no need
+                .adminPrivateKey(getJenkinsPrivateKey()) // temporary due to jclouds bug
+                .authorizeAdminPublicKey(true).adminPublicKey(getJenkinsPublicKey()).adminHome(getFsRoot()).build();
```

Comment:
Try replacing

getJenkinsUser()/getJenkinsPrivateKey()/getJenkinsPublicKey()

with

getAdminUser()/getCloud().getGlobalPrivateKey()/getCloud().getGlobalPublicKey()

Frankly, I'm not shure myself. My problems with this rather obscure
construct are:

1. org.jclouds.scriptbuilder.statements.login.AdminAccess.Builder is
completely undocumented.
2. My TestImages are Standard-Images and use cloud-init, so I'm *not*
able to test this part because this "initscript" feature did not even
work for me with the old code (before I started changes).

So I changed this part to the best of my knowledge, but could still be
wrong.

Cheers
 -Fritz

On 02/23/2015 02:42 PM, Oliver Gondža wrote:
> In
> jclouds-plugin/src/main/java/jenkins/plugins/jclouds/compute/JCloudsSlaveTemplate.java
> <https://github.com/jenkinsci/jclouds-plugin/pull/83#discussion_r25163584>:
> 
>>              AdminAccess adminAccess = AdminAccess.builder().adminUsername(getJenkinsUser())
>> -                    .installAdminPrivateKey(installPrivateKey) // some VCS such as Git use SSH authentication
>> -                    .grantSudoToAdminUser(allowSudo) // no need
>> -                    .adminPrivateKey(getCloud().privateKey) // temporary due to jclouds bug
>> -                    .authorizeAdminPublicKey(true).adminPublicKey(getCloud().publicKey).adminHome(getFsRoot()).build();
>> -
>> +                .installAdminPrivateKey(installPrivateKey) // some VCS such as Git use SSH authentication
>> +                .grantSudoToAdminUser(allowSudo) // no need
>> +                .adminPrivateKey(getJenkinsPrivateKey()) // temporary due to jclouds bug
>> +                .authorizeAdminPublicKey(true).adminPublicKey(getJenkinsPublicKey()).adminHome(getFsRoot()).build();
> 
> I am experimenting with your fix and I experience these when slaves are
> provisioned. You seem to run it with build user identity, not the admin
> identity.
> 
> |1) AuthorizationException on node <<LOCATION>>/<<INSTANCEID>>:
> org.jclouds.rest.AuthorizationException: (jenkins:rsa[fingerprint(<<FINGERPRINT>>),sha1(<<SHA1>>)]@<<IP>>) (jenkins:rsa[fingerprint(<<FINGERPRINT>>),sha1(<<SHA1>>)]@<<IP>>) error acquiring {hostAndPort=<<IP>>, loginUser=jenkins, ssh=null, connectTimeout=60000, sessionTimeout=60000} (out of retries - max 7): Exhausted available authentication methods
>   at org.jclouds.sshj.SshjSshClient.propagate(SshjSshClient.java:384)
>   at org.jclouds.sshj.SshjSshClient.acquire(SshjSshClient.java:204)
>   at org.jclouds.sshj.SshjSshClient.connect(SshjSshClient.java:224)
>   at org.jclouds.compute.callables.RunScriptOnNodeAsInitScriptUsingSsh.call(RunScriptOnNodeAsInitScriptUsingSsh.java:72)
>   at org.jclouds.compute.strategy.CustomizeNodeAndAddToGoodMapOrPutExceptionIntoBadMap.call(CustomizeNodeAndAddToGoodMapOrPutExceptionIntoBadMap.java:121)
>   at org.jclouds.compute.strategy.CustomizeNodeAndAddToGoodMapOrPutExceptionIntoBadMap.apply(CustomizeNodeAndAddToGoodMapOrPutExceptionIntoBadMap.java:142)
>   at org.jclouds.compute.strategy.CustomizeNodeAndAddToGoodMapOrPutExceptionIntoBadMap.apply(CustomizeNodeAndAddToGoodMapOrPutExceptionIntoBadMap.java:49)
>   at shaded.com.google.common.util.concurrent.Futures$2.apply(Futures.java:760)
>   at shaded.com.google.common.util.concurrent.Futures$ChainingListenableFuture.run(Futures.java:906)
>   at shaded.com.google.common.util.concurrent.Futures$1$1.run(Futures.java:635)
>   at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
>   at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
>   at java.lang.Thread.run(Thread.java:745)
> Caused by: net.schmizz.sshj.userauth.UserAuthException: Exhausted available authentication methods
>   at net.schmizz.sshj.userauth.UserAuthImpl.authenticate(UserAuthImpl.java:114)
>   at net.schmizz.sshj.SSHClient.auth(SSHClient.java:205)
>   at net.schmizz.sshj.SSHClient.authPublickey(SSHClient.java:305)
>   at net.schmizz.sshj.SSHClient.authPublickey(SSHClient.java:324)
>   at org.jclouds.sshj.SSHClientConnection.create(SSHClientConnection.java:165)
>   at org.jclouds.sshj.SSHClientConnection.create(SSHClientConnection.java:49)
>   at org.jclouds.sshj.SshjSshClient.acquire(SshjSshClient.java:194)
>   ... 11 more
> Caused by: net.schmizz.sshj.userauth.UserAuthException: publickey auth failed
>   at net.schmizz.sshj.userauth.UserAuthImpl.handle(UserAuthImpl.java:157)
>   at net.schmizz.sshj.transport.TransportImpl.handle(TransportImpl.java:474)
>   at net.schmizz.sshj.transport.Decoder.decode(Decoder.java:127)
>   at net.schmizz.sshj.transport.Decoder.received(Decoder.java:195)
>   at net.schmizz.sshj.transport.Reader.run(Reader.java:72)
> |
> 
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jenkinsci/jclouds-plugin/pull/83/files#r25163584>.
>